# t-тест


In [ ]:
import numpy as np
import pandas as pd
import statsmodels.formula.api as smf
from scipy.stats import t # t-distribution

## sleep equation
Значимость влияния `union`

контрольные переменные `totwrk, age, male`

Неробастный тест

In [ ]:
sleep_df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
mod1 = smf.ols(formula='sleep~union+totwrk+age+male', data=sleep_df).fit()
mod1.summary(slim=True)

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sleep   R-squared:                       0.117
Model:                            OLS   Adj. R-squared:                  0.112
No. Observations:                 706   F-statistic:                     23.14
Covariance Type:            nonrobust   Prob (F-statistic):           5.50e-18
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   3461.4144     68.709     50.378      0.000    3326.514    3596.315
union         10.2120     38.236      0.267      0.789     -64.859      85.283
totwrk        -0.1658      0.018     -9.205      0.000      -0.201      -0.130
age            2.7495      1.396      1.970      0.049       0.009       5.490
male          84.7408     34.424      2.462      0.014      17.155     152.327
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.02e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Имеем: $\hat{\beta}_{union}=10.2120$, s.e.($\beta$)=38.236, $t=\frac{\hat{\beta}}{s.e.(\beta)}=0.267$, $P_{union}=0.789$=78.9%

Уровень значимости $\alpha$=5%=0.05

Вывод: $P_{union}>\alpha$ => **коэфициент незначим!**, т.е. гипотеза о том, что `union` не влияет на `sleep`, не отвергается/согласуется с наблюдениями

In [ ]:
# Необходимое критическое значения, df= .df_resid
t.ppf(q=1-0.05/2, df=mod1.df_resid)

1.9633538601819327

$t_{cr}\approx1.96$, $t=0.267$, $|t|<t_{cr}$ => коэффициент незначим (t-статистика "маленькая")

## sleep equation 2

Рассморим модель `sleep` на `totwrk, totwrk^2` (объясняющие), `age, male, smsa, south` (контрольные)

Посмотрим на значимость `totwrk^2`

In [ ]:
sleep_df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
mod2 = smf.ols(formula='sleep~totwrk+I(totwrk**2)+age+male+smsa+south', data=sleep_df).fit()
print(mod2.summary(slim=True))

                            OLS Regression Results                            
Dep. Variable:                  sleep   R-squared:                       0.134
Model:                            OLS   Adj. R-squared:                  0.127
No. Observations:                 706   F-statistic:                     18.05
Covariance Type:            nonrobust   Prob (F-statistic):           1.58e-19
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       3414.1444     76.730     44.495      0.000    3263.495    3564.794
totwrk            -0.0940      0.048     -1.956      0.051      -0.188       0.000
I(totwrk ** 2) -1.899e-05   1.11e-05     -1.708      0.088   -4.08e-05    2.84e-06
age                2.8387      1.383      2.052      0.041       0.123       5.554
male              90.7097     34.155      2.656      0.008      23.651     157.769
smsa             -60.523

$\beta_{totwrk^2}\approx -1.899*10^{-5}=-0.00001899$, $P_{totwrk^2}=0.088=8.8\%$, $t_{totwrk^2}=-1.708$

Пусть уровень значимости 10%=0.1

In [ ]:
# критическое значение
t.ppf(q=1-0.10/2, df=mod2.df_resid)

1.6470364637535286

$t_{cr}\approx 1.65$

Имеем: $|t|>t_{cr}$, $P<\alpha$ => коэффициент значим! Зависимость сна от занятости нелинейная (непостоянное предельное значение!)

## sleep equation 2 (робастный тест)
Рассморим модель `sleep` на `totwrk, totwrk^2` (объясняющие), `age, male, smsa, south` (контрольные)

Посмотрим на значимость `totwrk^2` (с поправкой на гетероскедастичность)

In [ ]:
sleep_df = pd.read_csv('https://raw.githubusercontent.com/artamonoff/Econometrica/master/python-notebooks/data-csv/sleep75.csv')
mod3 = smf.ols(formula='sleep~totwrk+I(totwrk**2)+age+male+smsa+south', data=sleep_df).fit(cov_type='HC3')
print(mod3.summary(slim=True))

                            OLS Regression Results                            
Dep. Variable:                  sleep   R-squared:                       0.134
Model:                            OLS   Adj. R-squared:                  0.127
No. Observations:                 706   F-statistic:                     14.08
Covariance Type:                  HC3   Prob (F-statistic):           3.57e-15
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept       3414.1444     82.857     41.205      0.000    3251.747    3576.542
totwrk            -0.0940      0.063     -1.500      0.134      -0.217       0.029
I(totwrk ** 2) -1.899e-05   1.57e-05     -1.208      0.227   -4.98e-05    1.18e-05
age                2.8387      1.354      2.097      0.036       0.186       5.492
male              90.7097     35.513      2.554      0.011      21.106     160.314
smsa             -60.523

$robust-P_{totwrk^2}=0.227=22.7\%$, $robust-t_{totwrk^2}=-1.208$. Критическое значение то же

Вывод: коэффициент незначим!